flink-libraries/flink-table/src/main/java/org/apache/flink/table/runtime/DateTimeFunctions.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.runtime;
+
+import org.joda.time.format.DateTimeFormatter;
+import org.joda.time.format.DateTimeFormatterBuilder;
+
+/**
+ * Built-in scalar functions for date time related operations.
+ */
+public class DateTimeFunctions {
+	private static final int PIVOT_YEAR = 2020;
+
+	private static final ThreadLocalCache<String, DateTimeFormatter> DATETIME_FORMATTER_CACHE =
+		new ThreadLocalCache<String, DateTimeFormatter>(100) {
+			@Override
+			protected DateTimeFormatter getNewInstance(String format) {
+				return createDateTimeFormatter(format);
+			}
+		};
+
+	public static String dateFormat(long ts, String formatString) {
+		DateTimeFormatter formatter = DATETIME_FORMATTER_CACHE.get(formatString);
```

Comment:
Would it make sense to add a shortcut during code-gen if the pattern is a string literal (this should be the most common case) to avoid the lookup?